include/ibamr/WaveGenerationStrategy.h

```diff
@@ -0,0 +1,220 @@
+//
+//    * Redistributions of source code must retain the above copyright notice,
+//      this list of conditions and the following disclaimer.
+//
+//    * Redistributions in binary form must reproduce the above copyright
+//      notice, this list of conditions and the following disclaimer in the
+//      documentation and/or other materials provided with the distribution.
+//
+//    * Neither the name of The University of North Carolina nor the names of its
+//      contributors may be used to endorse or promote products derived from
+//      this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+/////////////////////// INCLUDE GUARD ////////////////////////////////////
+
+#ifndef included_WaveGenerationStrategy
+#define included_WaveGenerationStrategy
+
+///////////////////////////// INCLUDES ///////////////////////////////////
+#include <string>
+
+#include "Variable.h"
+#include "ibamr/AdvDiffHierarchyIntegrator.h"
+#include "ibamr/INSVCStaggeredHierarchyIntegrator.h"
+#include "ibtk/ibtk_utilities.h"
+#include "tbox/Pointer.h"
+
+namespace IBAMR
+{
+/*!
+ * A collection of post processing call back functions to be hooked into
+ * IBAMR::HierarchyIntegrator class to employ a wave generation zone of some
+ * prescribed length within the computational domain. Additional strategies
+ * can be implemented as simple C functions within this collection.
+ *
+ * The instance of a WaveGenerationStrategy and the particular generation function
+ * should be registered with an appropriate hierarchy integrator via
+ * registerPostprocessIntegrateHierarchyCallback.
+ *
+ * \param ctx is the pointer to WaveGenerationStrategy struct.
+ */
+
+namespace WaveGenerationFunctions
+{
+void callStokesWaveRelaxationCallbackFunction(double current_time,
+                                              double new_time,
+                                              bool skip_synchronize_new_state_data,
+                                              int num_cycles,
+                                              void* ctx);
+} // namespace WaveGenerationFunctions
+
+/*! Struct for generating waves based on relaxation method.
+ *
+ */
+struct WaveGenerationStrategy
```

Comment:
Technically this is not a "strategy" class (it does not do anything). Possibly it should be an abstract base class, or a base class with default constructors and destructors.

@amneetb, how general is this meant to be? Is this something that really only works for Stokes waves, or are there other things that you could imaging using this for?

Reply:
We will add more wave generation from other theories like: cnoidal, irregular etc as we go along. This should be easily extendable.

Reply:
OK. Should the virtual functions be moved into the base class (`WaveGenerationStrategy`)?

Reply:
We can, but I think not all wave theories can readily provide surfaceElevation() and/or velocity. Hence, I kept this to StokesWaveClass only. Stokes-[I-IV] can give provide this estimate easily. `WaveGenerationStrategy` is meant to hold required data structures that all of the wave generation strategies will need.

Reply:
OK, well, at present `WaveGenerationStrategy` is not really a [strategy](https://sourcemaking.com/design_patterns/strategy) -- so it should get a new name.

Reply:
Agreed. Is there a design pattern name where the base class provides just data structures, rather than methods? I used `strategy` in a sense to hint that it not useful directly, as it provides no functionality.

Reply:
Usually we want to use inheritance / class hierarchies to specify the interface in terms of a set of virtual functions. If there is not a common set of functions that will be implemented, my inclination is not to put the classes in the same hierarchy.